compiler/next/test/frontend/testParseConditional.cpp

```diff
@@ -0,0 +1,347 @@
+  assert(cond->thenStmt(0)->isComment());
+  assert(cond->thenStmt(0)->toComment()->str() == "/* c4 */");
+  assert(cond->thenStmt(1)->isFnCall());
+  assert(!cond->hasElseBlock());
+  assert(cond->numElseStmts() == 0);
+  assert(!cond->isExpressionLevel());
+
+  // Make sure the then iterator works as expected.
+  {
+    std::array<ASTTag, 2> thenStmtList = {
+      asttags::Comment,
+      asttags::FnCall
+    };
+    auto i = 0;
+    for (const auto stmt : cond->thenStmts()) {
+      assert(stmt->tag() == thenStmtList[i]);
+      assert(stmt->tag() == cond->thenStmt(i++)->tag());
+    }
+  }
+}
+
+static void test4(Parser* parser) {
+  auto parseResult = parser->parseString("test4.chpl",
+      "/* c1 */\n"
+      "if foo += bar() then\n"
+      "  baz();\n"
+      "/* c2 */\n");
+  assert(parseResult.errors.size() == 0);
+  assert(parseResult.topLevelExpressions.size() == 1);
+  assert(parseResult.topLevelExpressions[0]->isModule());
+  auto mod = parseResult.topLevelExpressions[0]->toModule();
+  assert(mod->numStmts() == 3);
+  assert(mod->stmt(0)->isComment());
+  assert(mod->stmt(1)->isConditional());
+  assert(mod->stmt(2)->isComment());
+  const Conditional* cond = mod->stmt(1)->toConditional();
+  assert(cond != nullptr);
+  assert(cond->condition());
+  assert(cond->condition()->isOpCall());
+  assert(!cond->condition()->toOpCall()->op().compare("+="));
+  assert(cond->thenBlockStyle() == BlockStyle::IMPLICIT);
+  assert(cond->numThenStmts() == 1);
+  assert(cond->thenStmt(0)->isFnCall());
+  assert(!cond->hasElseBlock());
+  assert(cond->numElseStmts() == 0);
+  assert(!cond->isExpressionLevel());
+}
+
+static void test5(Parser* parser) {
+  auto parseResult = parser->parseString("test5.chpl",
+      "/* c1 */\n"
+      "if var foo = bar() then\n"
+      "  baz();\n"
+      "/* c2 */\n");
+  assert(parseResult.errors.size() == 0);
+  assert(parseResult.topLevelExpressions.size() == 1);
+  assert(parseResult.topLevelExpressions[0]->isModule());
+  auto mod = parseResult.topLevelExpressions[0]->toModule();
+  assert(mod->numStmts() == 3);
+  assert(mod->stmt(0)->isComment());
+  assert(mod->stmt(1)->isConditional());
+  assert(mod->stmt(2)->isComment());
+  const Conditional* cond = mod->stmt(1)->toConditional();
+  assert(cond != nullptr);
+  assert(cond->condition());
+  const Variable* condVar = cond->condition()->toVariable();
+  assert(condVar);
+  assert(!condVar->name().compare("foo"));
+  assert(!condVar->typeExpression());
+  assert(condVar->initExpression());
+  assert(condVar->initExpression()->isFnCall());
+  assert(cond->thenBlockStyle() == BlockStyle::IMPLICIT);
+  assert(cond->numThenStmts() == 1);
+  assert(cond->thenStmt(0)->isFnCall());
+  assert(!cond->hasElseBlock());
+  assert(cond->numElseStmts() == 0);
+  assert(!cond->isExpressionLevel());
+}
+
+int main() {
+  Context context;
+  Context* ctx = &context;
+
+  auto parser = Parser::build(ctx);
+  Parser* p = parser.get();
+
+  test0(p);
```

Comment:
This is a broader issue than this particular PR is about, but has there been thought of if we should use GoogleTest or some other unit testing harness?

Reply:
Personally, I don't expect that will more than 100~ tests or so for the new compiler (might be a bit off base here but that's a rough guess). I expect that as we replace the old frontend with the new one that we will gradually begin to rely more and more on existing nightly testing.